esp-hal/src/gpio/mod.rs

```diff
@@ -2054,6 +1945,55 @@ impl<'lt> AnyPin<'lt> {
         });
     }
 
+    fn clear_interrupt(&self) {
+        self.bank().write_interrupt_status_clear(self.mask());
+    }
+
+    fn with_gpio_lock<F, R>(&self, f: F) -> R
+    where
+        F: FnOnce() -> R,
+    {
+        // If the pin is listening, we need to take a critical section to prevent racing
```

Comment:
I initially thought this might be racy, as another core/interrupt could call `listen`, but given that all these are `self` methods, and `listen` takes `&mut self`, I think that's impossible to do (without unsafe)?

Reply:
Normally there can be only one AnyPin, so this can only race with an interrupt handler. But if that's possible, we take GPIO_LOCK. At least this is the idea :)

Reply:
Actually I did introduce a way to race: Flex::split_into_drivers allows `Input::wait_for` and `Output::apply_config` at the same time. It would be easy to say that it is UB to do, but perhaps that API is just bad and not worth patching.